### PR TITLE
Staking eligibility enforced per court

### PR DESCRIPTION
**Pull request by jaybuidl**
Intended for the most specialized courts only. Disabled unless `KlerosCore.Court.eligibility` is set to other than `address(0)`. 

Many files are touched by this PR but it's mostly because of the extra parameter returned by the `courts()` getter. 

Previous attempts at tackling this issue: #2232, #1653 

## Overview

`ICourtEligibility` is a **pluggable per-court predicate** that Kleros uses to decide whether an address is allowed to **increase stake** (and, by extension, be in the juror pool for draws) for that court.

### What the interface is
`ICourtEligibility` is intentionally tiny: one view function.

```5:13:contracts/src/arbitration/interfaces/ICourtEligibility.sol
interface ICourtEligibility {
    /// @notice Checks if the juror is eligible to stake or to vote in the court.
    function isEligible(address _juror, uint96 _courtID) external view returns (bool);
}
```

Think of it as: “Given `juror` and `courtID`, should we treat them as eligible for that court’s juror set?”

### Where it’s stored (KlerosCore)
Each `Court` has an `eligibility` field pointing to a contract that implements `ICourtEligibility`.

```40:52:contracts/src/arbitration/KlerosCore.sol
struct Court {
    // ...
    ICourtEligibility eligibility; // The eligibility predicate for the court.
    // ...
}
```

It’s set/updated by governance when courts are created/modified:

- `createCourt(..., ICourtEligibility _eligibility)` sets `court.eligibility = _eligibility`
- `changeCourtParameters(..., ICourtEligibility _eligibility)` updates it

### Where it’s enforced (SortitionModule via KlerosCore)
The enforcement point is **staking validation** inside the SortitionModule, called from `KlerosCore._setStake()`.

In `SortitionModule._validateStake`, eligibility is only checked on **stake increases**:

```300:305:contracts/src/arbitration/SortitionModule.sol
if (stakeIncrease) {
    // Check if the juror is eligible to stake in the court.
    if (_eligibility != ICourtEligibility(address(0)) && !_eligibility.isEligible(_account, _courtID)) {
        return (0, 0, StakingResult.NotEligibleForStaking);
    }
    // ...
}
```

Then `KlerosCore` converts that staking result into a revert:

```1411:1453:contracts/src/arbitration/KlerosCore.sol
(uint256 pnkDeposit, uint256 pnkWithdrawal, StakingResult stakingResult) = sortitionModule.validateStake(
    _account,
    _courtID,
    _newStake,
    _noDelay,
    courts[_courtID].eligibility
);
// ...
if (_result == StakingResult.NotEligibleForStaking) revert NotEligibleForStaking();
```

So the runtime behavior is:

- If `court.eligibility == address(0)`: **no extra eligibility restriction** (everyone can stake, subject to other rules).
- Else: stake increases require `eligibility.isEligible(account, courtID) == true`, otherwise revert `NotEligibleForStaking`.

### How dispute kits relate
Dispute kits can *also* implement `ICourtEligibility`, but they don’t have to. When they do, it lets the court’s eligibility predicate be “whatever logic the dispute kit wants.”

Examples:

- **`DisputeKitGated` / `DisputeKitGatedShutter`** implement `isEligible` based on token-gate configuration for that `courtID` (ERC721 list + ERC1155 tokenId lists).
- **`DisputeKitSybilResistant`** implements `isEligible` based on Proof of Humanity, ignoring `_courtID`.

```71:74:contracts/src/arbitration/dispute-kits/DisputeKitSybilResistant.sol
function isEligible(address _juror, uint96 /* _courtID */) external view override returns (bool) {
    return poh.isHuman(_juror);
}
```

So the dispute kit is effectively being reused as a **court policy module** for staking eligibility.

### The mental model / flow
- **KlerosCore** owns court config (including the pointer to an eligibility contract).
- **SortitionModule** is the gatekeeper for staking changes and enforces eligibility on stake increases by calling `eligibility.isEligible`.
- **Dispute kits** are one possible implementation of that eligibility predicate (token gating, sybil resistance, etc.).

---

## PR-Codex overview
This PR primarily introduces the `ICourtEligibility` interface and its integration across various contracts and tests to enforce eligibility checks for jurors in staking processes.

### Detailed summary
- Added `ICourtEligibility` interface with `isEligible` function.
- Integrated eligibility checks in `SortitionModule` and `KlerosCore`.
- Updated dispute kits to implement eligibility logic.
- Modified tests to cover eligibility scenarios and edge cases.
- Renamed and adjusted existing tests to reflect new eligibility requirements.

> The following files were skipped due to too many changes: `contracts/test/arbitration/helpers/dispute-kit-gated-common.ts`, `contracts/src/arbitration/dispute-kits/DisputeKitGated.sol`, `contracts/src/arbitration/dispute-kits/DisputeKitGatedShutter.sol`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

## Summary by CodeRabbit

* **New Features**
  * Courts support per‑court eligibility predicates surfaced in events and enforced during staking and juror draws.
  * Token‑gated courts now support multiple ERC721/ERC1155 collections and token‑ID gating with query APIs.
  * Sybil‑resistant mode exposes human‑eligibility checks; staking can yield an explicit "not eligible" outcome.

* **Tests**
  * Test suites updated to provide and assert the new eligibility parameter and the expanded token‑gating behavior.

### CODE REVIEW

File: contracts/src/arbitration/dispute-kits/DisputeKitGated.sol [hunk 1]
```diff
@@ -92,20 +144,125 @@ contract DisputeKitGated is DisputeKitClassicBase {
...
+            EnumerableSet.UintSet storage tokenIds = supportedErc1155TokenIds[_courtID][token];
+            uint256 tokenIdsLength = tokenIds.length();
+            for (uint256 j = 0; j < tokenIdsLength; j++) {
+                uint256 tokenId = tokenIds.at(j);
+                if (IBalanceHolderERC1155(token).balanceOf(_juror, tokenId) > 0) return true;
+            }
+        }
+        return false;
+    }
+
```
Comment (jaybuidl) [inline contracts/src/arbitration/dispute-kits/DisputeKitGated.sol]: We need to add a bunch of views explicitly because the EnumerableSet mappings cannot be public due to the way OZ has implemented it (nested struct).

File: contracts/src/arbitration/dispute-kits/DisputeKitClassicBase.sol [hunk 2]
```diff
@@ -651,7 +651,7 @@ abstract contract DisputeKitClassicBase is IDisputeKit, Initializable, UUPSProxi
         Round storage round = dispute.rounds[dispute.rounds.length - 1];
 
         (uint96 courtID, , , , , ) = core.disputes(_coreDisputeID);
-        (, bool hiddenVotes, , , , ) = core.courts(courtID);
```
Comment (jaybuidl) [inline contracts/src/arbitration/dispute-kits/DisputeKitClassicBase.sol]: Most of the code changes in the diff are caused by this extra court parameter

**Review by salgozino — APPROVED** [review]
LGTM!

Disclaimer, I haven't reviewed that the court return params (`(currentCourtID, , , , , , ) = core.courts(currentCourtID); // Get the parent court.`) were changed everywhere :blush:
